25.Blur/main.cpp

```diff
@@ -4,182 +4,500 @@
+		assetManager = std::move(initOutput.assetManager);
+		cpu2gpuParams = std::move(initOutput.cpu2gpuParams);
+		logger = std::move(initOutput.logger);
+		inputSystem = std::move(initOutput.inputSystem);
+
+	}
+
+	void onAppTerminated_impl() override
+	{
+		logicalDevice->waitIdle();
+	}
+
+	void workLoopBody() override
+	{
+        inputSystem->getDefaultKeyboard(&keyboard);
+        keyboard.consumeEvents(
+            [this](const ui::IKeyboardEventChannel::range_t& events)
+            {
+                for (auto eventIt = events.begin(); eventIt != events.end(); ++eventIt)
+                {
+                    const auto& ev = *eventIt;
+                    if ((ev.keyCode == ui::EKC_Q) && (ev.action == ui::SKeyboardEvent::ECA_RELEASED))
+                    {
+                        m_appRunning = false;
+                    }
+                }
+            },
+            logger.get());
+
+        inputSystem->getDefaultMouse(&mouse);
+        mouse.consumeEvents([this](const ui::IMouseEventChannel::range_t& events)
+            {
+                for (auto eventIt = events.begin(); eventIt != events.end(); ++eventIt)
+                {
+                    const auto& ev = *eventIt;
+                    if (ev.type == ui::SMouseEvent::EET_SCROLL)
+                    {
+                        logger->log("Mouse vertical scroll event detected: %d\n", system::ILogger::ELL_DEBUG, ev.scrollEvent.verticalScroll);
+                    }
+                }
+            }, logger.get());
+	}
+
+	bool keepRunning() override
+	{
+		return m_appRunning && windowCb->isWindowOpen();
+	}
+
+private:
+	core::smart_refctd_ptr<nbl::ui::IWindowManager> windowManager;
+	core::smart_refctd_ptr<nbl::ui::IWindow> window;
+	core::smart_refctd_ptr<CommonAPI::CommonAPIEventCallback> windowCb;
+	core::smart_refctd_ptr<nbl::video::IAPIConnection> apiConnection;
+	core::smart_refctd_ptr<nbl::video::ISurface> surface;
+	core::smart_refctd_ptr<nbl::video::IUtilities> utilities;
+	core::smart_refctd_ptr<nbl::video::ILogicalDevice> logicalDevice;
+	video::IPhysicalDevice* physicalDevice;
+	std::array<video::IGPUQueue*, CommonAPI::InitOutput::MaxQueuesCount> queues;
+	core::smart_refctd_ptr<nbl::video::ISwapchain> swapchain;
+	core::smart_refctd_ptr<video::IGPURenderpass> renderpass = nullptr;
+	std::array<nbl::core::smart_refctd_ptr<video::IGPUFramebuffer>, CommonAPI::InitOutput::MaxSwapChainImageCount> fbos;
+	std::array<std::array<nbl::core::smart_refctd_ptr<nbl::video::IGPUCommandPool>, CommonAPI::InitOutput::MaxFramesInFlight>, CommonAPI::InitOutput::MaxQueuesCount> commandPools;
+	core::smart_refctd_ptr<nbl::system::ISystem> system;
+	core::smart_refctd_ptr<nbl::asset::IAssetManager> assetManager;
+	video::IGPUObjectFromAssetConverter::SParams cpu2gpuParams;
+	core::smart_refctd_ptr<nbl::system::ILogger> logger;
+	core::smart_refctd_ptr<CommonAPI::InputSystem> inputSystem;
+	video::IGPUObjectFromAssetConverter cpu2gpu;
```

Comment:
do you need all of this?

Reply:
Most of it yeah, probably not one or two things like `utilities` and perhaps `windowCb`. Will remove the excess stuff.